arangod/Pregel/Conductor/RecoveringState.cpp

```diff
@@ -0,0 +1,149 @@
+#include "RecoveringState.hpp"
+
+#include "Pregel/Algorithm.h"
+#include "Pregel/Conductor.h"
+#include "Pregel/Conductor/State.hpp"
+#include "Pregel/MasterContext.h"
+#include "Pregel/PregelFeature.h"
+#include "Pregel/Recovery.h"
+#include "Pregel/WorkerConductorMessages.h"
+
+using namespace arangodb::pregel::conductor;
+
+#define LOG_PREGEL(logId, level)          \
+  LOG_TOPIC(logId, level, Logger::PREGEL) \
+      << "[job " << conductor._executionNumber << "] "
+
+Recovering::Recovering(Conductor& conductor) : conductor{conductor} {
+  conductor.updateState(ExecutionState::RECOVERING);
+}
+
+Recovering::~Recovering() {}
+
+auto Recovering::run() -> void {
+  if (conductor._algorithm->supportsCompensation() == false) {
+    LOG_PREGEL("12e0e", ERR) << "Algorithm does not support recovery";
+    conductor.changeState(conductor::StateType::Canceled);
+    return;
+  }
+
+  // we lost a DBServer, we need to reconfigure all remainging servers
+  // so they load the data for the lost machine
+  conductor._statistics.reset();
+
+  TRI_ASSERT(SchedulerFeature::SCHEDULER != nullptr);
+
+  // let's wait for a final state in the cluster
```

Comment:
I'm really not sure why this has to be queued and cannot be directly executed here. Any ideas?
This queuing makes it quite hard: In line 60 we have to test again, if we are still in the Recovering state because due to stuff we did in the lines 37 to 59 we could have reached the Canceled state already. If we don't queue, we can be sure in which state we are.

Reply:
I have no immediate idea why this is the case.

I believe though that once we sorted out the state machine and interfaces to conductor and workers we will have an easier time determining which code should run asynchronously at which point we will sort these out with a fine comb.